text/3735-field-projections.md

```diff
@@ -0,0 +1,1744 @@
+itself and `OFFSET` is the offset in bytes as returned by `offset_of!`.
+
+
+With the above projections in place, users can write the following code:
+
+```rust
+struct Foo {
+    bar: i32,
+    baz: u32,
+}
+
+let x: &Foo = &Foo { bar: 42, baz: 43 };
+let y: &Foo = &Foo { bar: 24, baz: 25 };
+let d: DoubleRef<'_, Foo> = DoubleRef::new(x, y);
+let bars: DoubleRef<'_, i32> = d->bar;
+assert_eq!((42, 24), bars.read());
+```
+
+#### Containers
+
+The other kind of projections are that of *containers*, for example `UnsafeCell<T>` or
+`MaybeUninit<T>`. They are governed by the `#[projecting]` attribute.
+
+If we want to combine the two containers from the previous sections, we can do it in the following
+way:
+
+```rust
+#[projecting]
+#[repr(transparent)]
+pub struct Opaque<T> {
+    inner: UnsafeCell<MaybeUninit<T>>,
+}
+
+impl<T> Opaque<T> {
+    pub fn new(value: T) -> Self {
+        Self { inner: UnsafeCell::new(MaybeUninit::new(value)) }
+    }
+
+    pub fn uninit() -> Self {
+        Self { inner: UnsafeCell::new(MaybeUninit::uninit()) }
+    }
+
+    pub fn write(&mut self, value: T) {
+        self.inner.get_mut().write(value);
+    }
+}
+```
+
+Now `&Opaque<T>` can represent a reference that points at data from other languages, such as C.
+
+The `#[projecting]` attribute changes the way the field types are generated for this type. Instead
+of having a field type representing the `inner` field, this type will "project" through to the type
+`T`, inheriting all fields that `T` has. So if we consider the type `Foo` from above:
+
+```rust
+struct Foo {
+    bar: i32,
+    baz: u32,
+}
+```
+
+Then there are two field types for `Opaque<Foo>`:
+- one representing `bar` with `Base = Opaque<Foo>` and `Type = Opaque<i32>`
+- the other representing `baz` with `Base = Opaque<Foo>` and `Type = Opaque<u32>`
+
+So the both the base type and the type of the fields are wrapped with the container. Now users can
+write:
+
+```rust
+fn init(foo: &mut Opaque<Foo>) {
+    let bar: &mut Opaque<i32> = foo->bar;
+    bar.write(42);
+    foo->baz.write(24);
+}
+```
+
+
+## Impact of this Feature
+
+Overall this feature improves readability of code, because it replaces more complex to parse syntax
+with simpler syntax:
+- `&raw mut (*ptr).foo` is turned into `ptr->foo`
+- using `NonNull<T>` as a replacement for `*mut T` becomes a lot better when accessing fields,
+
+There is of course a cost associated with introducing a new operator along with the concept of field
+projections.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Implementation Details
+
+In order to facilitate field projections, several interlinked concepts have to be introduced. These
+concepts are:
+
+- [field types],
+  - `Field` traits,
+  - `field_of!` macro,
+  - [`#[projecting]`](#projecting-attribute) attribute
+- projection operator `->`,
+  - `Project` trait,
+  - `Projectable` trait,
```

Comment:
I'm trying to figure out what will eventually need to be stable so field projection can be used on library types, vs. what will need to be stable to make custom types field projectable. Is this list accurate?

- With access to `->`, a user can make use of basic field projection on library types and pointers
- With access to `#[projecting]` and/or the ability to implement `Project` and `Projectable`, a user can make their own types field projectable
- With access to the `Field` trait members (but not necessarily ability to implement) and `field_of!`, a user can express field projection in function signatures
- `Field` and `UnalignedField` may never be manually implemented, they are compiler-generated for any field that is accessed by `field_of!`

Reply:
> I'm trying to figure out what will eventually need to be stable so field projection can be used on library types, vs. what will need to be stable to make custom types field projectable. Is this list accurate?

It is, but one can add additional caveats/information:

> * With access to `->`, a user can make use of basic field projection on library types and pointers

`->` also gives access to projections for custom types (if we depend on another crate using unstable features to implement custom projections).

> * With access to `#[projecting]` and/or the ability to implement `Project` and `Projectable`, a user can make their own types field projectable

Yes, but often one needs the information contained in the `Field` trait members, since how else can one project to a field? (of course if the stdlib already provides the projection for raw pointers and one has a raw pointer then it's fine)

> * With access to the `Field` trait members (but not necessarily ability to implement) and `field_of!`, a user can express field projection in function signatures

One only needs the `Field` trait to express field projection in function signatures. Of course to call them, one needs the `field_of!` macro (or the syntax).

> * `Field` and `UnalignedField` may never be manually implemented, they are compiler-generated for any field that is accessed by `field_of!`

Yes, that is the current idea, but for some reasons we might want to lift that restriction, see [`Field` trait stability](https://github.com/y86-dev/rfcs/blob/field-projection-v2/text/3735-field-projections.md#field-trait-stability)

Reply:
i think an enum variant field implementing something other than `Field` (unless the enum has only 1 variant, in which case it's effectively both a struct and an enum) would work well.

Reply:
I'm not sure I understand correctly, what are you replying to in this thread? Do you mean as a way to implement enum projections?